aten/src/ATen/native/Sorting.cpp

```diff
@@ -94,6 +91,169 @@ void quick_select_template(
+
+  do {
+    do {
+      ++first;
+    } while (lt(*first, *pivot));
+    do {
+      --last;
+    } while (lt(*pivot, *last));
+    if (last < first) {
+      break;
+    }
+    std::swap(*first, *last);
+  } while (true);
+  std::swap(*pivot, *last);
+
+  return last;
+}
+
+template <typename random_it, typename kth_iterator, typename comp>
+void quickselect_helper(
+    random_it start,
+    random_it first,
+    random_it last,
+    kth_iterator kbegin,
+    kth_iterator kend,
+    comp lt) {
+  // Base case: one element or no kth values
+  if (last <= first || kbegin == kend) {
+    return;
+  }
+
+  // Base case: two elements
+  if (last == first + 1) {
+    if (lt(*last, *first)) {
+      std::swap(*first, *last);
+    }
+    return;
+  }
+
+  random_it pivot = partition(first, last, lt);
+  int64_t pivot_rank = std::distance(start, pivot);
+
+  auto lt_kend = std::lower_bound(kbegin, kend, pivot_rank);
+  quickselect_helper(start, first, pivot - 1, kbegin, lt_kend, lt);
+
+  auto gt_kbegin = std::upper_bound(lt_kend, kend, pivot_rank);
+  quickselect_helper(start, pivot + 1, last, gt_kbegin, kend, lt);
+}
+
+// Rearranges elements in range [first, last) such that kths elements
+// are in their sorted location.
+template <typename random_it, typename comp>
+void quickselect(
+    random_it first,
+    random_it last,
+    std::vector<int64_t> kths,
+    comp lt) {
+  return quickselect_helper(
+      first, first, last - 1, kths.begin(), kths.end(), lt);
+}
+
+void quantile_cpu_impl(
+    Tensor& out,
+    const Tensor& self,
+    const Tensor& q,
+    const optional<int64_t>& _dim,
+    bool ignore_nan) {
+  int64_t dim = at::maybe_wrap_dim(_dim.value_or(0), self.dim(), true);
+
+  // Checks that all q values are between 0 and 1, inclusive
+  // NOTE: this check is only performed when running on the CPU to avoid
+  // synchronizing an accelerator with the CPU
+  TORCH_CHECK(
+      q.ge(0).logical_and_(q.le(1)).all().item<bool>(),
+      "quantile() q values must be in the range [0, 1]");
+
+  // If dim is None flatten. Move dimension to reduce as last dim and flatten
+  // remaining dims so in shape is num_reductions by size of dim to reduce.
+  Tensor in;
+  if (!_dim || self.dim() <= 1) {
+    in = self.clone(at::MemoryFormat::Contiguous).view({1, -1});
+  } else if (dim == self.dim() - 1) {
+    in = self.clone(at::MemoryFormat::Contiguous).view({-1, self.size(-1)});
+  } else {
+    in = self.unsqueeze(-1)
+             .transpose_(dim, -1)
+             .clone(at::MemoryFormat::Contiguous)
+             .view({-1, self.size(dim)});
+  }
+
+  int64_t q_size = q.numel();
+  int64_t num_reductions = in.size(0);
+  int64_t reduction_size = in.size(1);
+
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "quantile_cpu_impl", [&] {
+    scalar_t* const OP = out.data_ptr<scalar_t>();
+    scalar_t* const IP = in.data_ptr<scalar_t>();
+    const scalar_t* const QP = q.data_ptr<scalar_t>();
+
+    auto lt = [](scalar_t a, scalar_t b) {
```

Comment:
note here about NaN ordering.

Could this be simplified as `(a < b) || _isnan(b)`? Putting the more frequent comparison first may shortcircuit the NaN checks, which is nice. Second, if there's a NaN then isn't it OK to say that A is always less than it in this case, even if A is NaN?